Introduce communication for projects and other resources

KochTobi opened the pull request:
![architecture-Project Resource Service (2)](https://user-images.githubusercontent.com/11536497/124269196-de5aa280-db3a-11eb-883e-cb5816cf548f.png)

**PR Checklist**
Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

- [x] This comment contains a description of changes (with reason)
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] `CHANGELOG.rst` is updated

**Description of changes**
This orients itself on the ResourcesService that was introduced int the offer manager.
However, it is improved in the regard that the information on what change was made is now also communicated.
Furthermore, the event emitter is not necessary and you only have one service/channel on which you publish changes.

KochTobi committed:
Introduce communication for projects and other resources

This orients itself on the ResourcesService that was introduced int he offer manager.
However, it is improved in the regards that the information on what change was made is now also communicated.
Furthermore the event emmitter is not necessary and you only have one service/channel on which you publish changes.

KochTobi committed:
Simplify generics and improve javadoc

KochTobi committed:
Update Changelog

KochTobi committed:
Merge branch 'development' into feature/project-service

KochTobi committed:
Remove unnecessary TODO

KochTobi committed:
Update Changelog

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/communication/Channel.groovy:
```diff
@@ -0,0 +1,46 @@
+package life.qbic.portal.sampletracking.communication
+/**
+ * A channel used to publish to a chosen topic and informs subscribers about publications
```
I am not sure if I understand that scentences.

```suggestion
 * A channel informs its subscribers about publications in the context of a given channel specific topic
```

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/communication/Message.groovy:
```diff
@@ -0,0 +1,9 @@
+package life.qbic.portal.sampletracking.communication
+
+/**
+ * <p>A message to be sent around. This class should be extended to hold relevant content.</p>
+ *
+ * @since 1.0.0
+ */
+abstract class Message {
+}
```
What's the benefit to use this class here? This class provides no functionality other than grouping classes. What was your idea in introducing it?

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/communication/Subscriber.groovy:
```diff
@@ -0,0 +1,15 @@
+
+/**
+ * The subscription interface serves as a functional interface
+ * to receive communication of type T emitted by a Channel.
```
```suggestion
 * to receive communication of type T emitted by a {@link Channel}.
```

KochTobi commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceMessage.groovy:
```diff
@@ -0,0 +1,28 @@
+/**
+ * <p>A message that communicates changes in a resource of type T</p>
+ *
+ * @param <T>  the data type stored in the resource
+ * @since 1.0.0
+ */
+abstract class ResourceMessage<T> extends Message {
+    public Optional<T> payload
+    private final MessageType type
+
+    ResourceMessage(T payload, MessageType type) {
```
We could also choose `V` here. This is the payload data type (for example `Project`)

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceMessage.groovy:
```diff
@@ -0,0 +1,28 @@
+ * @param <T>  the data type stored in the resource
+ * @since 1.0.0
+ */
+abstract class ResourceMessage<T> extends Message {
```
I dont see the benefit of Message here. It would be sufficient to know its a ResourceMessage

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,69 @@
+     * @param resourceItem the resource item to add
+     * @since 1.0.0
+     */
+    abstract <RESOURCE extends V> void addToResource(RESOURCE resourceItem)
```
Why is resource written here in capital letters?

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,69 @@
+     * @since 1.0.0
+     */
+    void clear() {
+        // we have to copy otherwise there is a concurrent modification exception
+        List<? extends V> toRemove = new ArrayList<>(content)
+        toRemove.forEach { removeFromResource(it) }
+    }
```
if we copy the content the new list will be empty but what is with the resource list? I would assume this will be still populated

KochTobi replied:
the new list is not touched. it is only used as a reference as to which elements to remove. the `toRemove` list is removed by the garbage collector after the `removeFromResource` was executed for all the elements. This method removes it then from the resource.

jenniferboedker replied:
ahh ok now I see it! :+1:

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,69 @@
+        toRemove.forEach { removeFromResource(it) }
+    }
+
+    /**
+     * This method clears the content of the resource service and populates it with a new set of data.
+     *
+     * @see #clear
```
? this does not fit the method signature below

KochTobi replied:
The clear() method should be used/ the behaviour is expected. That is why I added it using `@see`. Could be more verbose in the description though :+1:

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/project/ProjectResourceService.groovy:
```diff
@@ -0,0 +1,34 @@
+    <RESOURCE extends Project> void addToResource(RESOURCE resourceItem) {
+        content.add(resourceItem)
+        ProjectAddedMessage projectAddedMessage = new ProjectAddedMessage(resourceItem)
+        publish(projectAddedMessage)
```
How about, instead of creating message type objects, sending the message type through the publish method?

I see the issue that we create lots and lots of message objects. They do not provide other functionality than having the message type. And if I see it correctly you do not enforce the use of the message type directly. With doing it through the publishe method you would still know the message type and you have less overhead with creating new services

KochTobi replied:
I introduced Topics now to do that. Thank you for your valuable feedback. I think it helped a lot in simplifying this.

jenniferboedker reviewed (dismissed):
Thanks for your suggestion! I like your idea and added some comments. I think you could improve your implementation if you would remove the classes and methods which benefit was not so clear for me. This directs especially to the Message class and the different Message types you create. I would try to KISS

KochTobi committed:
Simplify service architecture

KochTobi committed:
Use generic parameter naming convention

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,113 @@
+ * @since 1.0.0
+ */
+abstract class ResourceService<T> {
+    protected final List<? extends T> content
```
👏🏽 upper boundary

jenniferboedker commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,113 @@
+     * @param subscriber The subscriber to register for update events
+     * @param topic the topic to subscribe to
+     */
+    void subscribe(Subscriber<T> subscriber, Topic topic) {
+        if( channels.containsKey(topic)) {
+            channels.get(topic).subscribe(subscriber)
+        } else {
+            Channel<T> topicChannel = new Channel<T>()
+            channels.put(topic, topicChannel)
+            topicChannel.subscribe(subscriber)
+        }
```
I would assume that the subscriber cannot define which topics are implemented.
Therefore I would expect there to be a method to add new topics

KochTobi replied:
Yes you are right. We need to decide on what to do then. We could throw an IllegalArgument exception.

KochTobi replied:
57d525f

KochTobi committed:
Remove possibility to add topic to a service

KochTobi committed:
Use RessourceService method to add topics

KochTobi committed:
Add missing since tags

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/communication/Channel.groovy:
```diff
@@ -0,0 +1,45 @@
+     * @param message the message to be sent
+     * @since 1.0.0
+     */
+    void publish(T message) {
+        for (Subscriber<T> subscriber : subscribers) {
+            subscriber.receive(message)
+        }
+    }
```
I think it is a good idea to introduce interfaces here. The subscribers should not be able to publish messages. The publisher can subscribe though, nothing speaks against that.

Or is it by design, that every client can publish and receive?

KochTobi replied:
I did not want to restrict this as long as we do not require it. The publish method is not reachable when using the resource services and I am not sure where the channel might be used apart from that. I think it is fine that everyone (being a subscriber or not) is able to publish.

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/communication/Subscriber.groovy:
```diff
@@ -0,0 +1,16 @@
+     * @param data the data that is received
+     */
+    void receive(T data)
+}
```
```suggestion
}

```

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,122 @@
+     * @param data the data to be published
+     * @param topic the topic to publish to
+     */
+    protected void publish(T data, Topic topic) {
+        channels.get(topic)?.publish(data)
+    }
```
I think it is quite crucial that the publisher can assure, that the topic exists. Otherwise the publisher what assume everything is fine.

An exception would be one possible solution.

KochTobi replied:
Will do :+1:

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,122 @@
+        channels.put(topic, new Channel<T>())
+    }
+
+    /**
+     * Adds a resource item to a resource of the service.
+     *
+     * @param resourceItem the resource item to add
+     * @since 1.0.0
+     */
+    abstract void addToResource(T resourceItem)
```
Since you make it abstract, you have to provide an example how the implementation is supposed to work (broadcasts the resource item to all subscribers of the "added" channel for example

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,122 @@
+     * @param resourceItem the resource item to remove
+     * @since 1.0.0
+     */
+    abstract void removeFromResource(T resourceItem)
```
same request as my previous comment

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,122 @@
+        return new ArrayList<>(this.content).iterator()
+    }
+
+    /**
+     * Clears the content of the resource service
+     *
+     * @since 1.0.0
+     */
+    void clear() {
+        // we have to copy otherwise there is a concurrent modification exception
+        List<? extends T> toRemove = new ArrayList<>(content)
+        toRemove.forEach { removeFromResource(it) }
+    }
```
if you call clear, i expect the channel for removing a resource is triggered.

KochTobi replied:
It is using the `removeFromResource` method. This way we do not need to implement it twice and trigger the same sequence

KochTobi replied:
I renamed the list to make it clearer.

sven1103 commented on sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy:
```diff
@@ -0,0 +1,122 @@
+     * @see #clear
+     * @since 1.0.0
+     */
+    void items(List<? extends T> items) {
+        clear()
+        for (def item in items) {
+            addToResource(item)
+        }
+    }
```
I don't understand the intent of this method

KochTobi replied:
This is basically a reset functionality. Like for vaadin grids this method makes sure that after it is called only the elements that are passed to it are contained in the resource. it could be used when populating a resource service in the dependency manager or use case, without having to care about the current contents, adding multiple items.

sven1103 replied:
is it needed for the current Sprint task? :) If not i vote for not implementing it, as we should not make assumption about the future (the future is fuzzy)

KochTobi committed:
Add newline at the end

Co-authored-by: Sven F. <sven.fillinger@qbic.uni-tuebingen.de>

KochTobi committed:
Avoid silent failure in publish method

Avoids the publish method from doing nothing.
If a topic does not exist it throws an illegal argument exception.

KochTobi committed:
Avoid silent failure for unsubscription

KochTobi committed:
Improve JavaDoc

KochTobi committed:
Rename variables an setItem method

KochTobi committed:
Remove unused methods

sven1103 reviewed (approved):
Very nice work!
